java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/common/PostgresValidator.java

```diff
@@ -212,25 +211,40 @@ private void validateTableSchema() throws SQLException {
             stmt.setString(2, this.tableName);
             var res = stmt.executeQuery();
 
-            // Field names in lower case -> data type
-            Map<String, String> schema = new HashMap<>();
+            class ColumnInfo {
+                String dataType;
+                Long charMaxLength;
+                String udtName;
+
+                ColumnInfo(String dataType, Long charMaxLength, String udtName) {
+                    this.dataType = dataType;
+                    this.charMaxLength = charMaxLength;
+                    this.udtName = udtName;
+                }
+            }
+
+            Map<String, ColumnInfo> schema = new HashMap<>();
             while (res.next()) {
                 var field = res.getString(1);
                 var dataType = res.getString(2);
-                schema.put(field.toLowerCase(), dataType);
+                Long charMaxLength =
+                        res.getObject(3) == null ? null : ((Number) res.getObject(3)).longValue();
+                var udtName = res.getString(4);
+                schema.put(field.toLowerCase(), new ColumnInfo(dataType, charMaxLength, udtName));
             }
 
             for (var e : tableSchema.getColumnTypes().entrySet()) {
                 // skip validate internal columns
                 if (e.getKey().startsWith(ValidatorUtils.INTERNAL_COLUMN_PREFIX)) {
                     continue;
                 }
-                var dataType = schema.get(e.getKey().toLowerCase());
-                if (dataType == null) {
+                var colInfo = schema.get(e.getKey().toLowerCase());
+                if (colInfo == null) {
                     throw ValidatorUtils.invalidArgument(
                             "Column '" + e.getKey() + "' not found in the upstream database");
                 }
-                if (!isDataTypeCompatible(dataType, e.getValue())) {
+                if (!isDataTypeCompatible(
+                        colInfo.dataType, e.getValue(), colInfo.charMaxLength, colInfo.udtName)) {
```

Comment:
nit: why not pass `colInfo` as a whole, in case there are more details in the future?